Dockerfile

```diff
@@ -1,29 +1,55 @@
-FROM golang:1.13 as secrets-provider-builder
+# =================== BASE BUILD LAYER ===================
+# this layer is used to prepare a common layer for both debug and release builds
+FROM golang:1.13 as secrets-provider-builder-base
 MAINTAINER CyberArk Software Ltd.
 
 ENV GOOS=linux \
     GOARCH=amd64 \
     CGO_ENABLED=0
 
-COPY . /opt/secrets-provider-for-k8s
+RUN apt-get update
 
-WORKDIR /opt/secrets-provider-for-k8s
+RUN go get -u github.com/jstemmer/go-junit-report && \
+    go get github.com/smartystreets/goconvey
+
+WORKDIR /opt/secrets-provider
 
 EXPOSE 8080
 
-RUN apt-get update && apt-get install -y jq
+COPY go.mod .
+COPY go.sum .
 
-RUN go get -u github.com/jstemmer/go-junit-report && \
-    go get github.com/smartystreets/goconvey
+# Add a layer of prefetched modules so in case of rebuild the modules are already cached
+RUN go mod download
+
+# =================== RELEASE BUILD LAYER ===================
+# this layer is used to build the release binaries
+FROM secrets-provider-builder-base as secrets-provider-builder
+
+COPY . .
```

Comment:
looks like this is done for both build layers (release and debug). can it move to the base build layer?

Reply:
I prefer not to.
This step copies the entire folder, and I want it to be the last one so previous steps could be cached (e.g. in debug build)

Reply:
ok thanks for clarifying.